# Discovery

A tour through the database — tables, sizes, indexes, and row counts. This is where we case the joint before the forensic analysis in notebook 02.

## Entity Relationship Diagram
![Entity Relationship Diagram](./autogenerated/er.png)

The schema is a Twitter/X clone: users, posts, comments, friends (bi-directional), plus time-series tables (metrics, events, logs) for performance experiments.

## Entity Relationship Diagram with Indexes
> notice gray cells

![Entity Relationship Diagram with Indexes](./autogenerated/indexes.png)

## 1. Connectivity Check

Verify the connection works.

In [1]:
%load_ext sql

Loading configurations from /app/pyproject.toml.

Settings changed:

Config,value
feedback,True
autopandas,False


In [2]:
%%sql
SELECT 1 + 1

1 rows affected.

?column?
2


## 2. Row Counts

How much data are we dealing with? These are the exact counts (not estimates) for the social tables.

In [3]:
%%sql
SELECT
    (SELECT COUNT(*) FROM users_profile) AS users_profile_count,
    (SELECT COUNT(*) FROM posts) AS post_count,
    (SELECT COUNT(*) FROM comments) AS comments_count,
    (SELECT COUNT(*) FROM friends) AS friends_count;

Running query in 'postgresql://postgres:***@postgres:5432/mydb'

1 rows affected.

users_profile_count,post_count,comments_count,friends_count
10000,150000,1200000,9999


## 3. Table & Index Sizes

Where does the storage go? The key metric here is `index_to_table_pct` — an index larger than the table itself should raise eyebrows.

In [4]:
%%sql
SELECT 
    relname AS table_name,
    pg_size_pretty(pg_relation_size(relid)) AS table_size,
    pg_size_pretty(pg_indexes_size(relid)) AS index_size,
    pg_size_pretty(pg_total_relation_size(relid)) AS total_size,
    ROUND(
        pg_indexes_size(relid)::numeric / NULLIF(pg_relation_size(relid), 0) * 100, 
        2
    ) AS index_to_table_pct
FROM pg_catalog.pg_statio_user_tables
ORDER BY pg_total_relation_size(relid) DESC;

Running query in 'postgresql://postgres:***@postgres:5432/mydb'

7 rows affected.

table_name,table_size,index_size,total_size,index_to_table_pct
metrics,607 MB,695 MB,1302 MB,114.58
logs,846 MB,299 MB,1146 MB,35.37
events,651 MB,391 MB,1043 MB,60.11
comments,147 MB,53 MB,200 MB,36.26
posts,21 MB,5560 kB,27 MB,25.30
users_profile,840 kB,1360 kB,2240 kB,161.90
friends,512 kB,648 kB,1192 kB,126.56


> **Observations**
> - `metrics` has **more index than table** (115%), largely from `idx_metric_time` on `(metric_name, recorded_at)` — a composite index on half the columns.
> - `users_profile` and `friends` show 162% and 127% respectively because their tables are tiny (kB-scale) while carrying PKs, constraint indexes, and FK indexes — the overhead ratio looks alarming but the absolute cost is small.
> - The time-series tables (metrics, logs, events) dominate storage at multiple GB each.

### Per-index breakdown

Which individual indexes are the heaviest, and are they actually used? `idx_scan` tells us how many index scans PostgreSQL has performed since the last stats reset.

In [5]:
%%sql
SELECT 
    schemaname,
    relname AS table_name,
    indexrelname AS index_name,
    idx_scan,
    pg_size_pretty(pg_relation_size(indexrelid)) AS index_size
FROM pg_stat_user_indexes
ORDER BY pg_relation_size(indexrelid) DESC;

Running query in 'postgresql://postgres:***@postgres:5432/mydb'

17 rows affected.

schemaname,table_name,index_name,idx_scan,index_size
public,metrics,idx_metric_time,0,481 MB
public,events,events_pkey,0,391 MB
public,logs,logs_pkey,0,214 MB
public,metrics,metrics_pkey,0,214 MB
public,logs,idx_session,0,85 MB
public,comments,comments_pkey,0,26 MB
public,comments,idx_comments_fk_post_id,0,18 MB
public,comments,idx_comments_fk_user_id,2,9584 kB
public,posts,posts_pkey,1200000,3312 kB
public,posts,idx_posts_fk_user_id,0,1288 kB


> **Red flags**
> - `events_pkey` (391 MB), `metrics_pkey` (214 MB), `logs_pkey` (214 MB), `comments_pkey` (26 MB) — **4 PK indexes with 0 scans**. Combined cost: **~845 MB** of dead weight.
> - `idx_metric_time` — **481 MB, 0 scans**. A composite index on `(metric_name, recorded_at)` that has never been used.
> - `idx_session` — **85 MB, 0 scans**. Index on logs.session_id with zero usage.
> - `idx_comments_fk_post_id` — **18 MB, 0 scans**. Unused FK index on the largest social table.
> - `idx_posts_fk_user_id` — **1,288 kB, 0 scans**. Unused FK index.
> - `idx_comments_fk_user_id` — **9,584 kB, only 2 scans**. Effectively unused.
>
> Also notable: `logs` shows `estimated_rows = -1`, meaning `ANALYZE` has never been run on it. All `last_analyze` columns show `None` — autovacuum/autoanalyze hasn't kicked in yet for this fresh database.

### Index-to-table size ratio per index

How much of each table's storage is eaten by each index?

In [6]:
%%sql
SELECT 
    s.schemaname,
    s.relname AS table_name,
    s.indexrelname AS index_name,
    s.idx_scan,
    pg_relation_size(s.indexrelid) AS index_size,
    pg_relation_size(t.relid) AS table_size,
    ROUND(
        pg_relation_size(s.indexrelid)::numeric / 
        NULLIF(pg_relation_size(t.relid), 0) * 100, 
        2
    ) AS pct_of_table
FROM pg_stat_user_indexes s
JOIN pg_stat_user_tables t 
    ON s.relid = t.relid
ORDER BY pct_of_table DESC;

Running query in 'postgresql://postgres:***@postgres:5432/mydb'

17 rows affected.

schemaname,table_name,index_name,idx_scan,index_size,table_size,pct_of_table
public,metrics,idx_metric_time,0,504340480,636198912,79.27
public,users_profile,users_profile_email_key,10000,573440,860160,66.67
public,users_profile,idx_users_profile_email,0,573440,860160,66.67
public,friends,friends_pkey,249975,335872,524288,64.06
public,events,events_pkey,0,410329088,682672128,60.11
public,metrics,metrics_pkey,0,224632832,636198912,35.31
public,friends,idx_friends_friend_id,0,180224,524288,34.38
public,users_profile,users_profile_pkey,1369999,245760,860160,28.57
public,friends,idx_friends_user_id,0,147456,524288,28.13
public,logs,logs_pkey,0,224632832,887267328,25.32


> **Duplicate index alert** — `users_profile` has two indexes on email:
> - `idx_users_profile_email` — 0 scans, 560 kB
> - `users_profile_email_key` — 10,000 scans, 560 kB
>
> The UNIQUE constraint (`users_profile_email_key`) already covers lookups by email. `idx_users_profile_email` is a **redundant index** — it exists separately and has never been used. Dropping it would save ~560 kB on every `users_profile` row operation.

## 4. Estimated Rows vs. Table Size

PostgreSQL's `pg_class.reltuples` gives us the planner's row estimates. Comparing these with `ANALYZE` timestamps reveals if stats are stale.

In [7]:
%%sql
SELECT 
    s.schemaname,
    s.relname,
    c.reltuples::bigint AS estimated_rows,
    s.last_analyze,
    pg_size_pretty(pg_total_relation_size(s.relid)) AS total_size
FROM pg_catalog.pg_stat_user_tables s
JOIN pg_catalog.pg_class c
    ON c.oid = s.relid
ORDER BY pg_total_relation_size(s.relid) DESC;

Running query in 'postgresql://postgres:***@postgres:5432/mydb'

7 rows affected.

schemaname,relname,estimated_rows,last_analyze,total_size
public,metrics,10000334,None,1302 MB
public,logs,-1,None,1146 MB
public,events,9999858,None,1043 MB
public,comments,1200000,None,200 MB
public,posts,150000,None,27 MB
public,users_profile,10000,None,2240 kB
public,friends,9999,None,1192 kB


## 5. Full Storage Summary

A clean overview of every user table, its row count, data size, index size, and total footprint.

In [8]:
%%sql
SELECT
    table_name,
    pg_size_pretty(pg_table_size(format('%I.%I', table_schema, table_name))) AS table_size,
    pg_size_pretty(pg_indexes_size(format('%I.%I', table_schema, table_name))) AS indexes_size,
    pg_size_pretty(pg_total_relation_size(format('%I.%I', table_schema, table_name))) AS total_size
FROM
    information_schema.tables
WHERE
    table_schema NOT IN ('pg_catalog', 'information_schema')
    AND table_type = 'BASE TABLE'
ORDER BY
    pg_total_relation_size(format('%I.%I', table_schema, table_name)) DESC;

Running query in 'postgresql://postgres:***@postgres:5432/mydb'

7 rows affected.

table_name,table_size,indexes_size,total_size
metrics,607 MB,695 MB,1302 MB
logs,846 MB,299 MB,1146 MB
events,651 MB,391 MB,1043 MB
comments,147 MB,53 MB,200 MB
posts,22 MB,5560 kB,27 MB
users_profile,880 kB,1360 kB,2240 kB
friends,544 kB,648 kB,1192 kB


## Key Takeaways

| Finding | Impact | Action (notebook) |
|---|---|---|
| 4 PK indexes + `idx_metric_time` + `idx_session` with 0 scans (~1.4 GB total) | Wasted storage + write overhead on every INSERT | Investigate query patterns → **02_indexes** |
| `idx_comments_fk_post_id` (18 MB, 0 scans) | Bloated FK index on the largest social table | Consider dropping or replacing → **02_indexes** |
| `idx_users_profile_email` duplicates UNIQUE constraint | Redundant write overhead | Drop redundant index → **02_indexes** |
| `metrics` indexes larger than table (115%) | Excessive index overhead on time-series data | Possibly replace with partial/covering indexes → **02_indexes** |
| `logs` has `estimated_rows = -1`, no `ANALYZE` ever run | Planner has no stats — may pick terrible plans | Run `ANALYZE` or wait for autovacuum |
| Time-series tables dominate storage (metrics, logs, events) | ~3.5 GB footprint could benefit from partitioning | **06_partitioning** |